File: src/Cpp/17_Slowniki/Zad10.cpp
#include <algorithm>
#include <cassert>
#include <string>
#include <unordered_map>
#include <vector>

//Otrzymujesz tekst w postaci napisu. Znajdz czestosc wystepowania kazdego ze slow w tekscie.
//Zignoruj roznice miedzy malymi i wielkimi literami. Pamietaj, slowa skladaja sie wylacznie z liter!

void wyczysc(std::string& napis)
{
    auto it = napis.begin();

    while (it != napis.end()) {
        if (ispunct(*it))
            napis.erase(it);
        else
            it++;
    }
}

void naMale(std::string& slowo)
{
    transform(slowo.begin(), slowo.end(), slowo.begin(), ::tolower);
}

std::vector<std::string> rozdzielSlowa(const std::string& napis)
{
    std::vector<std::string> wynik;
    int pocz = 0;
    int konc = 0;
    while ((konc = napis.find(' ', pocz)) != std::string::npos) {
        if (konc != pocz) {
            auto slowo = napis.substr(pocz, konc - pocz);
            wyczysc(slowo);
            naMale(slowo);
            if (!slowo.empty())
                wynik.push_back(slowo);
        }
        pocz = konc + 1;
    }
    if (konc != pocz) {
        auto slowo = napis.substr(pocz);
        wyczysc(slowo);
        naMale(slowo);
        if (!slowo.empty())
            wynik.push_back(slowo);
    }

    return wynik;
}

std::unordered_map<char, int> budujSlownik(const std::string& slowo)
{
    std::unordered_map<char, int> slownik;

    for (const auto& znak : slowo)
        slownik[znak]++;

    return slownik;
}

std::vector<std::vector<std::string>> znajdzAnagramy(const std::string& tekst)
{

    auto slowa = rozdzielSlowa(tekst);

    std::vector<std::vector<std::string>> wynik;

    auto it = slowa.begin();

    while (it != slowa.end()) {
        auto it2 = it;
        auto slowo = *it;
        auto histogram = budujSlownik(slowo);

        std::vector<std::string> anagramy{ slowo };

        it2++;
        while (it2 != slowa.end()) {
            slowo = *it2;

            if (histogram == budujSlownik(slowo)) {
                if (find(anagramy.begin(), anagramy.end(), slowo) == anagramy.end())
                    anagramy.push_back(slowo);
                it2 = slowa.erase(it2);
            } else
                it2++;
        }

        if (anagramy.size() > 1)
            wynik.push_back(anagramy);

        it = slowa.erase(it);
    }

    return wynik;
}

void test1()
{
    std::string slowo = "To absurd, ze tyran Brudas, ten straszliwy bandzior sprawuje rzady w tym kraju. "
                        "Burza nad galeria i alergia na narty to zadna zbrodnia, jak bandzior i jego arbuz.";

    std::vector<std::vector<std::string>> wynik{ { "absurd", "brudas" }, { "tyran", "narty" }, { "bandzior", "zbrodnia" }, { "burza", "arbuz" }, { "galeria", "alergia" } };

    auto anagramy = znajdzAnagramy(slowo);
    assert(equal(anagramy.begin(), anagramy.end(), wynik.begin()));
}

void test2()
{
    std::string slowo = "I love the way sometimes the letters of an anagram rearrange themselves in my "
                        "head with no effort, whereas sometimes I have to write them out in a circle to work it out.";

    std::vector<std::vector<std::string>> wynik;

    auto anagramy = znajdzAnagramy(slowo);
    assert(equal(anagramy.begin(), anagramy.end(), wynik.begin()));
}

int main()
{

    test1();
    test2();

    return 0;
}

File: src/python/17_slowniki/zad10.py
"""
Tytul: Anagramy w tekscie
Otrzymujesz tekst w postaci napisu. Twoim zadaniem jest znalezienie wszystkich slow, w ktorych te same znaki wystepuja te sama liczbe razy. Innymi slowy, szukasz anagramow w tekscie. Nalezy zignorowac roznice miedzy malymi i wielkimi literami. Pamietaj, ze slowa skladaja sie wylacznie z liter.
Dane wejsciowe: Napis.
Dane wyjsciowe: Lista list napisow.
Przyklad:
Dla otrzymanego napisu:
"To absurd, ze tyran Brudas, ten straszliwy bandzior sprawuje rzady w tym kraju. Burza nad galeria i alergia na narty to zadna zbrodnia jak bandzior i jego arbuz.",
powinna zostac zwrocona lista:
[["absurd", "brudas"], ["tyran", "narty"], ["bandzior", "zbrodnia"], ["burza", "arbuz"], ["galeria", "alergia"]].

"""

import string

def podziel_zdanie_na_slowa(zdanie):
    """
    Funkcja zwraca liste slow ze zdania.
    """
    return zdanie.translate(str.maketrans("", "", string.punctuation)).split()

def na_male_litery(slowa):
    """
    Funkcja zamienia wielkie litery ze slow z listy slowa na male litery.
    """
    return [slowo.lower() for slowo in slowa]

def histogram(napis):
    """
    Funkcja zwraca slownik zawierajacy wszystkie litery w napisie oraz czestosc ich wystepowania.
    """
    histogram = {}
    for znak in napis:
        if znak in histogram:
            histogram[znak] += 1
        else:
            histogram[znak] = 1
    return histogram

def znajdz_wszystkie_anagramy_w_tekscie(napis):
    """
    Funkcja zwraca liste wszystkich anagramow w napisie.
    """
    slowa = podziel_zdanie_na_slowa(napis)
    slowa = na_male_litery(slowa)
    slowa = list(set(slowa))

    wynik = []
    i = 0
    while i < len(slowa):
        slowo = slowa[i]
        histogram_dla_slowa = histogram(slowo)
        wynik.append([slowo])
        slowa.remove(slowo)
        j = 0
        while j < len(slowa):
            inne_slowo = slowa[j]
            if inne_slowo != slowo:
                histogram_dla_innego_slowa = histogram(inne_slowo)
                if histogram_dla_slowa == histogram_dla_innego_slowa:
                    wynik[-1].append(inne_slowo)

            j += 1

        if len(wynik[-1]) == 1:
            wynik.pop()

        i += 1

    return wynik

def czy_listy_list_rowne(lista_a, lista_b):
    """
    Funkcja zwraca True jesli lista_a i lista_b skladaja sie z list, ktore maja taka sama ilosc elementow i elementy w tych listach sa takie same.
    """
    _lista_a = [tuple(sorted(x)) for x in lista_a]
    _lista_b = [tuple(sorted(x)) for x in lista_b]

    return set(_lista_a) == set(_lista_b)

def test_znajdz_wszystkie_anagramy_w_tekscie():
    assert znajdz_wszystkie_anagramy_w_tekscie("Ala ma kota") == []
    assert czy_listy_list_rowne(
        znajdz_wszystkie_anagramy_w_tekscie(
            "To absurd, ze tyran Brudas, ten straszliwy bandzior sprawuje rzady w tym kraju. Burza nad galeria i alergia na narty to zadna zbrodnia, jak bandzior i jego arbuz."
        ),
        [
            ["absurd", "brudas"],
            ["tyran", "narty"],
            ["bandzior", "zbrodnia"],
            ["burza", "arbuz"],
            ["galeria", "alergia"],
        ],
    )

if __name__ == "__main__":

    test_znajdz_wszystkie_anagramy_w_tekscie()

